### ISSUE TITLE
Question: Known behavior when workers become unresponsive when processing E01 files? 

### ISSUE THREAD

**Issue by rodgermoore** [log2timeline/plaso#631]
VMware Workstation 12 on a WIndows 10 host system.
Guest is Ubuntu 14.04 with latest updates.
4 CPU threads, 4 Gigs mem.
Plaso 1.4

I made an E01 image of my Win10 256 SSD using Guymager (Encase 6 format). It contains 3 VSS copies. When I run log2timeline directly on the E01 image, processing only the Windows 10 partition and all VSS copies, workers hangs after a while and get killed finishing with errors.

When I first mount the E01 using ewfmount $ewfmount /path/to/file.E01 /mnt/ewf
Plaso runs just fine on /mnt/ewf/ewf1

When I monitor workers using htop it seems Plaso is consuming more memory when processing the E01 directly. Also, when I force Plaso to use only 1 worker it seems stable too.

Is this known behavior? Do I need to post details?

**Comment by joachimmetz** [comment]
> When I monitor workers using htop it seems Plaso is consuming more memory when processing the E01 directly. 

Correct it will have the E01 overhead in the main and every worker process, with ewfmount it will have the overhead only once. Also ewfmount is able to limit the number of open file handles plaso/dfvfs currently is not.

However this does not explain why the workers end up in a non-responsive state. Anything in the logs that indicate workers failing?

**Comment by rodgermoore** [comment]
Processing /mnt/ewf/ewf1 directly got log2timeline killed too, it just took way more time. Syslog gave the answer:

Mar  8 00:06:48 u1404-plaso-ts kernel: [109760.801252] Out of memory: Kill process 5404 (log2timeline.py) score 208 or sacrifice child

Mar  8 00:06:48 u1404-plaso-ts kernel: [109760.801269] Killed process 5404 (log2timeline.py) total-vm:1283236kB, anon-rss:864084kB, file-rss:0kB

Increased memory size to 8Gig on VM and running log2timeline again.

**Comment by joachimmetz** [comment]
Note that a large E01 requires a fair bit of memory for overhead. Switch to RAW or QCOW2 if you want less run-time overhead.

**Comment by rodgermoore** [comment]
It seems game folders (Steam, Elite Dangerous) in the E01 are causing trouble. These folders have huge amounts (100.000+) of small and large zip, texture, exe, ... files. Memory at 8Gigs and increasing workers to 8 magically made Plaso process them successfully. Does this make any sense?

**Comment by joachimmetz** [comment]
Which file system?

**Comment by rodgermoore** [comment]
> Which file system?
Windows 10 so NTFS

**Comment by joachimmetz** [comment]
Ack could have been a FAT-variant as well, I asume that the file system contains the Steam, Elite Dangerous is NTFS based on "I made an E01 image of my Win10 256 SSD".

Something I'll have to take a closer look is what libtsk/pytsk does with a directory with that vast amount of entries.

**Comment by rodgermoore** [comment]
> I asume that the file system contains the Steam, Elite Dangerous is NTFS based on "I made an E01 image of my Win10 256 SSD".

It is.

> Something I'll have to take a closer look is what libtsk/pytsk does with a directory with that vast amount of entries.

Ok. Don't know if it's a common use-case though...

**Comment by rodgermoore** [comment]
> Something I'll have to take a closer look is what libtsk/pytsk does with a directory with that vast amount of entries.

BTW. If you want to test and need a good example. Install DCS World and DCS World 2 by Eagle Dynamics. Its free and has a biljion gazillion (zipped)files. http://www.digitalcombatsimulator.com/en/index.php

If you start with a low amount of workers (2) I think you will be able to reproduce the issue.

**Comment by joachimmetz** [comment]
Thx for the tip. Hopefully that helps in making a test image that allows me to reproduce the issue.

> If you start with a low amount of workers (2) I think you will be able to reproduce the issue.

I personally was thinking about doing some isolated tests with libtsk/pysk first.

**Comment by rodgermoore** [comment]
> I personally was thinking about doing some isolated tests with libtsk/pysk first.
You're the expert :wink: 
Good luck

**Comment by rodgermoore** [comment]
Still doing some testing on the same image. It seems VSS is also important to mention. If I disable VSS, there are 3 in this image, succesratio is higher too. It's becoming quite complex pinpointing the exact cause of the issue.

**Comment by joachimmetz** [comment]
Ack. Likely a combination of the memory consumption of the different layers interacting with one another. The more reason to baseline the behaviour per layer first.

**Comment by rodgermoore** [comment]
Yesterday I ran into yet another E01 image causing Plaso to choke. And again it was a Windows image packed with gaming installations and 1 VSS. This time I was able to notice exactly where workers stopped working. 
````
/Users/USERNAME/AppData/Local/Steam/htmlcache/* 
````
This directory is filled with gzip zipped cache files. When Plaso hits this directory all CPU threads run 100%.

The status window shows all workers and the column "File" shows GZIP:/Users/USERNAME/AppData/Local/Steam/htmlcache/f_000* 

Eventually Plaso stops working with the message "Processing completed with errors".

I am processing the E01 using ewfmount so using /mnt/ewf/ew1
